lib/tqueue.ml

```diff
@@ -0,0 +1,65 @@
+type 'a t = {
+  queue : 'a Queue.t;
+  lock : Lwt_mutex.t;
+  has_elt : unit Lwt_condition.t;
+}
+
+let create () =
+  let queue = Queue.create () in
+  let lock = Lwt_mutex.create () in
+  let has_elt = Lwt_condition.create () in
+  { queue; lock; has_elt }
+
+let add x { queue; lock; has_elt } =
+  Lwt_mutex.with_lock lock (fun () ->
+      Queue.add x queue;
+      let%lwt () =
+        if Queue.length queue = 1 then (
+          Lwt_condition.signal has_elt ();
+          Lwt.return ())
+        else
+          Lwt.return () in
+      Lwt.return ())
+
+let push = add
+
+let take { queue; lock; _ } =
+  Lwt_mutex.with_lock lock (fun () -> Lwt.return (Queue.take queue))
+
+let take_opt { queue; lock; _ } =
+  Lwt_mutex.with_lock lock (fun () -> Lwt.return (Queue.take_opt queue))
+
+let wait_to_take { queue; lock; has_elt } =
+  Lwt_mutex.with_lock lock (fun () ->
+      let%lwt () =
+        if Queue.is_empty queue then
+          Lwt_condition.wait ~mutex:lock has_elt
+        else
+          Lwt.return () in
+      Lwt.return (Queue.take queue))
+
+let pop = take
+
+let peek { queue; lock; _ } =
```

Comment:
I think it is common to provide these two functions, but what about removing this one and only use the safe one?

Reply:
Similar to above comment, I'm basically just copying [OCaml's Queue module](https://ocaml.org/api/Queue.html) but making it thread-safe and adding the functionality with condition vars to wait for an element to appear

Reply:
Ok got it, I think we should get rid of the unsafe functions.